0000-friends_in_high_kindednesses.md

```diff
@@ -39,231 +39,412 @@ rather than for as long as the borrow initiated by `next`.
+below makes it unambiguous and easy to see which parameters remain undetermined
+at the point of assigning the associated type constructor to a concrete type
+constructor.
+
+When used in a type position, of course, all of the parameters to an associated
+type constructor must have been applied to concrete types or type parameters that
+are in scope, so that it can be evaluated to a proper type.
+
+## Extending HRTBs
+
+Providing bounds on the types constructed by associated type constructors
+requires heavy use of HRTBs, or higher-ranked trait bounds. This exists in Rust
+today, but in a limited form, and it is an obscure feature primarily used in
+the background to make function traits behave as expected.
+
+In brief, a higher-ranked trait bound is one in which a type or lifetime
+parameter is introduced only for the scope of that trait bound. A classic
+example of how this can be useful is in the contrast between these two
+functions:
 
 ```rust
-trait StreamingIterator {
-   type Item<'a>;
-   fn next(&mut self) -> Option<Self::Item>;
+fn foo1<T, F>(x: T, id: F) -> T where F: Fn(T) -> T {
+    id(x)
+}
+
+fn foo2<T, F>(x: T, id: F) -> T where F: for<X> Fn(X) -> X {
+    id(x)
+}
+
+// Valid (evaluates to 4)
+foo1::<i32, _>(2, |x| x + x)
+
+// Invalid (type error)
+foo2::<i32, _>(2, |x| x + x)
+```
+
+In the second function, we _guarantee_ that the `id` argument is the identity
+function (ignoring side effects), because it must be a valid function of `X`
+to `X` for _all_ `X`, whereas the first can be specialized to only be a valid
+function for the type `T`, in this case `i32`.
+
+Higher-ranked trait bounds have several other use cases. Currently, Rust uses
+them to declare that arguments with different lifetimes can be passed to
+function types, by requiring that that function be valid for all lifetimes,
+rather than just for some single lifetime parameter.
+
+In order to bound associated type constructors, we use higher-ranked types to
+require that the type constructor constructs type which meet some bound. This
+can be done both in the declaration and when bounding a type parameter by a
+trait, and can be both a trait bound and a type equality bound. Here are
+examples in code, with their meanings written out:
+
+```rust
+trait Sequence<T> {
+    // For every lifetime, this constructor applied to that lifetime must
+    // produce a type which is an iterator of references of that lifetime
+    type Iter<'a> where for<'a> Iter<'a>: Iterator<Item=&'a T>;
+}
+
+// For every lifetime, the associated type constructor Item applied to
+// that lifetime produces a reference of that lifetime to a slice of bytes.
+struct Foo<T> where T: for<'a> StreamingIterator<Item=&'a mut [u8]> {
+   ...
 }
 ```
 
+Enabling this requires extending HRTBs in two different ways:
+
+1. First, allowing HRTBs to applied to type constructors on the left-hand side
+of the bound. Currently they are only available to traits on the right-hand
```

Comment:
[I thought we](https://github.com/rust-lang/rfcs/pull/1598#issuecomment-215996572) determined [otherwise](http://is.gd/EUeLBB). Or do you mean something outside the form of `for<'x> T: U`? If something different it's probably worth clarifying.